torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,120 @@
+import functools
+
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import autograd_not_implemented
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+__all__ = ["trace_wrapped"]
+
+
+# Trace wrapped is a higher order op meant for both invoking a bound function,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the functions as needed. This, in turn, means we can support functions in backward with complex python
+# state mutation. If we were to not do this, the functions would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
+def trace_wrapped(*args, fn):
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+_trace_wrapped_op = HigherOrderOperator("trace_wrapped")
+
+
+def _self_invoke(*args, fn):
+    # This wrapper intercepts calls to fn, and calls the real fn via _trace_wrapped
+    # Dynamo unpacks this higher order op into calling the wrapped fn
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+def _assert_meta(grad, size, stride, dtype):
+    assert grad.size() == size, "size mismatch"
+    assert grad.stride() == stride, "stride mismatch"
+    assert grad.dtype == dtype, "dtype mismatch"
+    return grad
+
+
+@_trace_wrapped_op.py_impl(ProxyTorchDispatchMode)
+def inner_trace(mode, *args, fn):
+    import torch
+
+    assert len(args) == 1
+    grad = args[0]
+    assert isinstance(grad, torch.Tensor)
+
+    if isinstance(fn, functools.partial):
+        fn.__name__ = fn.func.__name__  # type: ignore[attr-defined]
+    # We've implemented a higher-order operator that remains consistent in proxy tensor tracing.
+    # However, Dynamo is aware and traces this into its genuine functionality.
+    # The operator's purpose is to facilitate invoking non-traceable functions
+    # and embedding them directly into the graph. Essentially, this transforms function
+    # calls into "leaf modules" as per traditional FX terminology.
+    # Note: Instead of naming it "allow_in_graph", we opted for a different name since "allow_in_graph"
+    # might imply that it's traceable, whereas this function is intrinsically non-traceable.
+    # Note2: I hate this name
+    fn = functools.partial(_self_invoke, fn=fn)
+    fn.__name__ = fn.func.__name__
+
+    proxy_args = pytree.tree_map(mode.tracer.unwrap_proxy, (grad,))
```

Comment:
This tree_map seems totally unnecessary? like, proxy_args is literally `(mode.tracer.unwrap_proxy(grad),)`

Reply:
Yea!